File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/user/usecase/GetMyValorantStoreHistoryUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.user.usecase;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.ports.AuthenticatedUserProvider;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantSkinRepository;
import fr.huiitre.tools.modules.riot.valorant.application.user.ports.ValorantStoreHistoryRepository;
import fr.huiitre.tools.modules.riot.valorant.application.user.view.ValorantStoreHistoryView;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class GetMyValorantStoreHistoryUseCase implements SecuredUseCase {

    private final AuthenticatedUserProvider authenticatedUserProvider;
    private final ValorantStoreHistoryRepository storeHistoryRepository;
    private final ValorantSkinRepository skinRepository;

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public GetMyValorantStoreHistoryUseCase(
            AuthenticatedUserProvider authenticatedUserProvider,
            ValorantStoreHistoryRepository storeHistoryRepository,
            ValorantSkinRepository skinRepository) {
        this.authenticatedUserProvider = authenticatedUserProvider;
        this.storeHistoryRepository = storeHistoryRepository;
        this.skinRepository = skinRepository;
    }

    public List<ValorantStoreHistoryView> execute() {
        Long userId = authenticatedUserProvider.getUserId();
        return storeHistoryRepository.findAllRawByUserId(userId).entrySet().stream()
                .map(entry -> new ValorantStoreHistoryView(
                        entry.getKey(),
                        entry.getValue().stream()
                                .map(skinId -> skinRepository.findById(skinId, userId))
                                .filter(Optional::isPresent)
                                .map(Optional::get)
                                .collect(Collectors.toList())
                ))
                .collect(Collectors.toList());
    }
}

File: api/Modules/Riot/Valorant/Application/User/Usecases/GetMyValorantStoreHistoryUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;
using Tools.Api.Modules.Riot.Valorant.Application.User.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.User.Views;

namespace Tools.Api.Modules.Riot.Valorant.Application.User.Usecases;

public sealed class GetMyValorantStoreHistoryUseCase(
    UseCaseAuthorizer authorizer,
    IValorantAuthRepository valorantAuthRepository,
    IValorantStoreHistoryRepository storeHistoryRepository,
    IValorantSkinRepository skinRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.ReadOnly;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task<List<ValorantStoreHistoryView>> Execute(long accountId)
    {
        if (!await valorantAuthRepository.ExistsByIdAndUserId(accountId, CurrentUser.UserId))
        {
            throw AppException.NotFound(
                "VALORANT_ACCOUNT_NOT_FOUND",
                "Ce compte Valorant est introuvable.");
        }

        var skinIdsByDate = await storeHistoryRepository.FindAllRawByAccountId(accountId);
        var history = new List<ValorantStoreHistoryView>();

        // TODO: une requête par skin, comme le Java. À regrouper si l'historique s'allonge.
        foreach (var (date, skinIds) in skinIdsByDate)
        {
            var skins = new List<ValorantSkinView>();

            foreach (var skinId in skinIds)
            {
                // Un skin disparu du catalogue est simplement omis de sa journée.
                if (await skinRepository.FindById(skinId, accountId) is { } skin)
                {
                    skins.Add(skin);
                }
            }

            history.Add(new ValorantStoreHistoryView(date, skins));
        }

        return history;
    }
}
